components/num-util/src/fixed_decimal.rs

```diff
@@ -0,0 +1,517 @@
+    ///
+    /// let dec: FixedDecimal = UnsignedWrapper(750usize).try_into().unwrap();
+    /// assert_eq!("750", dec.to_string());
+    /// ```
+    fn try_from(value: UnsignedWrapper<T>) -> Result<Self, Self::Error> {
+        FixedDecimal::from_ascending(UintIterator::from(value.0))
+    }
+}
+
+impl FixedDecimal {
+    /// Initialize a FixedDecimal with an iterator of digits in ascending
+    /// order of magnitude, starting with the digit at magnitude 0.
+    ///
+    /// This method is not public; use TryFrom::<UintIterator<T>> instead.
+    fn from_ascending<T>(digits_iter: T) -> Result<Self, Error>
+    where
+        T: Iterator<Item = u8>,
+    {
+        // TODO: make X a usize generic to customize the size of this array
+        // https://github.com/rust-lang/rust/issues/44580
+        const X: usize = 32;
+        // A temporary structure to allow the digits in the iterator to be reversed.
+        // The digits are inserted started from the end, and then a slice is copied
+        // into its final destination (result.digits).
+        let mut mem: [u8; X] = [0u8; X];
+        let mut trailing_zeros: usize = 0;
+        let mut i: usize = 0;
+        for (x, d) in digits_iter.enumerate() {
+            // Take only up to std::i16::MAX values so that we have enough capacity
+            if x > std::i16::MAX as usize {
+                return Err(Error::Limit);
+            }
+            // TODO: Should we check here that `d` is between 0 and 9?
+            // That should always be the case if UintIterator is used.
+            if i != 0 || d != 0 {
+                i += 1;
+                match X.checked_sub(i) {
+                    Some(v) => mem[v] = d,
+                    // This error should be obsolete after X is made generic
+                    None => return Err(Error::Limit),
+                }
+            } else {
+                trailing_zeros += 1;
+            }
+        }
+        let mut result: Self = Default::default();
+        if i != 0 {
+            let magnitude = trailing_zeros + i - 1;
+            debug_assert_le!(magnitude, std::i16::MAX as usize);
+            result.magnitude = magnitude as i16;
+            result.upper_magnitude = result.magnitude;
+            debug_assert_le!(i, X);
+            result.digits.extend_from_slice(&mem[(X - i)..]);
+        }
+        debug_assert!(result.check_invariants());
+        return Ok(result);
+    }
+
+    /// Gets the digit at the specified order of magnitude. Returns 0 if the magnitude is out of
+    /// range of the currently visible digits.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu_num_util::FixedDecimal;
+    ///
+    /// let dec = FixedDecimal::from(945);
+    /// assert_eq!(0, dec.digit_at(-1));
+    /// assert_eq!(5, dec.digit_at(0));
+    /// assert_eq!(4, dec.digit_at(1));
+    /// assert_eq!(9, dec.digit_at(2));
+    /// assert_eq!(0, dec.digit_at(3));
+    /// ```
+    pub fn digit_at(&self, magnitude: i16) -> u8 {
```

Comment:
Why doesn't it return an `Option`?

Reply:
There are 5 "zones" to consider:

1. magnitude is smaller than lowest_magnitude
2. magnitude is larger than lowest_magnitude but smaller than the smallest nonzero digit
3. magnitude is a nonzero digit
4. magnitude is smaller than upper_magnitude but larger than the largest nonzero digit
5. magnitude is larger than upper_magnitude

Zones 2, 3, and 4 are the visible digits declared by the FixedDecimal settings.  Zones 1 and 5 are out of bounds and could be considered None.  However, I do not see this as a useful distinction to make.  First, zero is a sensible and correct fallback.  Second, if the user wants to know the bounds, they should use the `magnitude_range` function.  I therefore leaned on the side of making this function as efficient as possible.

Reply:
It feels fishy to me. I never encountered an API in Rust that would return `0` or some default value from a call that asks for an item at a position that is not available.
@Manishearth - how does it sound to you?

Reply:
It's not that the digit is "not available" (as stated above, ones 2 and 4 are definitely expected to fall back to 0); it's just that if the digit did exist, it would be a 0.  If you have the number 123, what is the digit in the thousands place?  In the tenths place?  They are both 0, even though they are out of range of the number 123.

Reply:
I see. Yes, that makes sense. Thank you for explaining this.